misk-aws/src/main/kotlin/misk/jobqueue/sqs/SqsJobConsumer.kt

```diff
@@ -53,70 +49,65 @@ internal class SqsJobConsumer @Inject internal constructor(
+      val messages = queue.call { client ->
+        client.receiveMessage(ReceiveMessageRequest()
+            .withAttributeNames("All")
+            .withMessageAttributeNames("All")
+            .withQueueUrl(queue.url)
+            .withMaxNumberOfMessages(10))
+            .messages
+      }
 
-            tracer.traceWithSpan("handle-job-${queueName.value}") { span ->
-              // If the incoming job has an original trace id, set that as a tag on the new span.
-              // We don't turn that into the parent of the current span because that would
-              // incorrectly include the execution time of the job in the execution time of the
-              // action that triggered the job
-              message.attributes[SqsJob.ORIGINAL_TRACE_ID_ATTR]?.let {
-                ORIGINAL_TRACE_ID_TAG.set(span, it)
-              }
+      val failure = AtomicBoolean(false)
+      messages.map { SqsJob(queueName, queues, metrics, it) }.forEach { message ->
+        dispatchThreadPool.submit {
+          metrics.jobsReceived.labels(queueName.value).inc()
+
+          tracer.traceWithSpan("handle-job-${queueName.value}") { span ->
+            // If the incoming job has an original trace id, set that as a tag on the new span.
+            // We don't turn that into the parent of the current span because that would
+            // incorrectly include the execution time of the job in the execution time of the
+            // action that triggered the job        q
+            message.attributes[SqsJob.ORIGINAL_TRACE_ID_ATTR]?.let {
+              ORIGINAL_TRACE_ID_TAG.set(span, it)
+            }
 
-              // Run the handler and record timing
-              try {
-                val (duration, _) = timed { handler.handleJob(message) }
-                metrics.handlerDispatchTime.record(duration.toMillis().toDouble(), queueName.value)
-              } catch (th: Throwable) {
-                log.error(th) { "error handling job from ${queueName.value}" }
-                metrics.handlerFailures.labels(queueName.value).inc()
-                Tags.ERROR.set(span, true)
-              }
+            // Run the handler and record timing
+            try {
+              val (duration, _) = timed { handler.handleJob(message) }
+              metrics.handlerDispatchTime.record(duration.toMillis().toDouble(), queueName.value)
+            } catch (th: Throwable) {
+              failure.set(true)
+              log.error(th) { "error handling job from ${queueName.value}" }
+              metrics.handlerFailures.labels(queueName.value).inc()
+              Tags.ERROR.set(span, true)
             }
           }
         }
       }
-    }
 
-    override fun close() {
-      if (!subscriptions.remove(queueName, this)) return
-
-      log.info { "closing subscription to queue ${queueName.value}" }
-      running.set(false)
+      return when {
+        failure.get() -> Status.FAILED
```

Comment:
don't you need to wait for the dispatched tasks to complete before checking this?

Reply:
umm yea. should probably have a test :(

Reply:
fixed in 9cdece6a298d316f06d8cbd6a2d8920ce4526c44

Reply:
added a test too